spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy

```diff
@@ -0,0 +1,1380 @@
+    MAVEN = "maven"
+}
+
+structure ResourcesParams {
+    @required
+    targets: BuildTargetIdentifiers
+}
+
+structure ResourcesResult {
+    @required
+    items: ResourceItems
+}
+
+structure ResourceItem {
+    @required
+    target: BuildTargetIdentifier
+    @required
+    /// List of resource files.
+    resources: URIs
+}
+
+list ResourceItems {
+    member: ResourceItem
+}
+
+structure OutputPathsParams {
+    @required
+    targets: BuildTargetIdentifiers
+}
+
+structure OutputPathsResult {
+    @required
+    items: OutputPathsItems
+}
+
+structure OutputPathsItem {
+    /// A build target to which output paths item belongs.
+    @required
+    target: BuildTargetIdentifier
+    /// Output paths.
+    @required
+    outputPaths: OutputPathItems
+}
+
+list OutputPathsItems {
+    member: OutputPathsItem
+}
+
+structure OutputPathItem {
+    /// Either a file or a directory. A directory entry must end with a forward
+    /// slash "/" and a directory entry implies that every nested path within the
+    /// directory belongs to this output item.
+    @required
+    uri: URI
+    /// Type of file of the output item, such as whether it is file or directory.
+    @required
+    kind: OutputPathItemKind
+}
+
+list OutputPathItems {
+    member: OutputPathItem
+}
+
+@enumKind("open")
+intEnum OutputPathItemKind {
+    /// The output path item references a normal file.
+    FILE = 1
+    /// The output path item references a directory.
+    DIRECTORY = 2
+}
+
+/// Task progress notifications may contain an arbitrary interface in their `data`
+/// field. The kind of interface that is contained in a notification must be
+/// specified in the `dataKind` field.
+///
+/// There are predefined kinds of objects for test and compile tasks, as described
+/// in the [Compile Request](#compile-request) and [Test Request]
+@enumKind("open")
+@dataFieldDiscriminator
+enum TaskDataKind {
+    /// `data` field must contain a CompileTask object.
+    COMPILE_TASK = "compile-task"
+    /// `data` field must contain a CompileReport object.
+    COMPILE_REPORT = "compile-report"
+    /// `data` field must contain a TestTask object.
+    TEST_TASK = "test-task"
+    /// `data` field must contain a TestReport object.
+    TEST_REPORT = "test-report"
+    /// `data` field must contain a TestStart object.
+    TEST_START = "test-start"
+    /// `data` field must contain a TestFinish object.
+    TEST_FINISH = "test-finish"
+}
+
+structure TaskStartParams {
+  /// Unique id of the task with optional reference to parent task id
+  @required
+  taskId: TaskId
+
+  /// Timestamp of when the event started in milliseconds since Epoch.
+  eventTime: EventTime,
+
+  /// Message describing the task.
+  message: String
+
+  /// Kind of data to expect in the `data` field. If this field is not set, the kind of data is not specified.
+  /// Kind names for specific tasks like compile, test, etc are specified in the protocol.
+  dataKind: TaskDataKind
```

Comment:
Out of the scope of this pr, but just a general thought. I wish we could restrict the type of `data` here based of `TaskDataKind`.

Reply:
This idea of a value restricting what `data` can be seems pretty common, but we just don't capture that in a structured way which is a bummer.

Reply:
See my response [there](https://github.com/build-server-protocol/build-server-protocol/pull/426#discussion_r1070608894)